c_binding/libses.c

```diff
@@ -0,0 +1,699 @@
+    }
+
+    memcpy(status, status_p, _T10_SES_DEV_SLOT_STATUS_LEN);
+
+    *gen_code_be = st_hdr->gen_code_be;
+
+ out:
+    if (rc != LSM_ERR_OK) {
+        memset(status, 0, _T10_SES_DEV_SLOT_STATUS_LEN);
+        *gen_code_be = 0;
+    }
+
+    return rc;
+}
+
+static int _ses_ctrl_data_gen(char *err_msg, uint8_t *status_data,
+                              uint8_t *status, const int16_t element_index,
+                              uint16_t *len)
+{
+    struct _ses_ctrl_diag_hdr *ctrl_hdr = NULL;
+    uint8_t *tmp_p = NULL;
+    uint8_t *end_p = NULL;
+
+    ctrl_hdr = (struct _ses_ctrl_diag_hdr *) (status_data);
+
+    *len = be16toh(ctrl_hdr->len_be) + 4;
+
+    /* set all element as not selected */
+    tmp_p = &ctrl_hdr->ctrl_dp_list_begin;
+    end_p = tmp_p + be16toh(ctrl_hdr->len_be);
+
+    while(tmp_p < end_p) {
+        _clear_array_bit(tmp_p, _T10_SES_CTRL_SELECT_BYTES,
+                         _T10_SES_CTRL_SELECT_BIT);
+        tmp_p += _T10_SES_DEV_SLOT_STATUS_LEN;
+    }
+
+    /* update the selected element */
+
+    tmp_p = &ctrl_hdr->ctrl_dp_list_begin + \
+            _T10_SES_DEV_SLOT_STATUS_LEN * element_index;
+
+    memcpy(tmp_p, status, _T10_SES_DEV_SLOT_STATUS_LEN);
+
+    return LSM_ERR_OK;
+}
+
+/*
+ * Workflow:
+ *  Parse config page(0x01)
+ *      Loop enclosure descriptor list
+ *          Loop descriptor header list
+ *              Store 'NUMBER OF POSSIBLE ELEMENTS' of descriptor header in
+ *              an array. The 0 indicate this element type only have overall
+ *              element.
+ */
+static int16_t _ses_eiioe(uint8_t *cfg_data, int16_t element_index)
+{
+    struct _ses_cfg_hdr *cfg_hdr = NULL;
+    struct _ses_cfg_dp_hdr *dp_hdr = NULL;
+    uint8_t *end_p = NULL;
+    uint8_t i = 0;
+    uint16_t total_dp_hdr_count = 0;
+    uint8_t add = 0;
+    uint8_t *dp_hdr_begin = NULL;
+
+    cfg_hdr = (struct _ses_cfg_hdr *)cfg_data;
+    end_p = cfg_data + be16toh(cfg_hdr->len_be) + 3;
+    if (end_p >= cfg_data + _SG_T10_SPC_RECV_DIAG_MAX_LEN)
+        /* Facing data boundary */
+        return -1;
+
+    _ses_cfg_parse(cfg_data, &dp_hdr_begin, &total_dp_hdr_count);
+    if ((dp_hdr_begin == NULL) || (total_dp_hdr_count == 0))
+        return -1;
+
+    for (i = 0; i < total_dp_hdr_count; ++i) {
+        dp_hdr = (struct _ses_cfg_dp_hdr *) dp_hdr_begin +
+            (_T10_SES_CFG_DP_HDR_LEN * i);
+        if ((uint8_t *) &dp_hdr >= end_p)
+            /* Facing data boundary */
+            return -1;
+        add++;
+        if (element_index <= dp_hdr->num_of_possible_element)
+            break;
+        element_index -= dp_hdr->num_of_possible_element;
+    }
+
+    return element_index + add;
+}
+
+/*
+ * Workflow:
+ *  1. Find out all enclosure /dev/sgX.
+ *  2. Find out which /dev/sgX given sas address is connecting to via this
+ *     SES pages:
+ *      6.1.13 Additional Element Status diagnostic page
+ *  3. Record down the element index of port given sas address is connecting to.
```

Comment:
Record the element index of the port that the given SAS address is connecting to.